Api user contact management

AngelinaMartineau opened the pull request.

AngelinaMartineau committed:
Add API contact management data model

- Modified API user class to contain e-mail subscriptions.
- Added migration script to add apiuser subscription table.
- Changed select api users query to use subqueries for roles and
 subscriptions
- Changed row mapper to handle new query results
- Changed SqlApiUserDao function to use row mapper instead of row
handler.
- Added add and remove functions for subscriptions
- Added unit and integration tests for ApiUser and Dao

refs #12330

AngelinaMartineau committed:
Modify Api User Service to handle user subscriptions

- change registerNewUser to add subscriptions for a user upon
registration
- add functions to get, add, and remove subscriptions in
ApiUserService.java
- create NewUserView.java to handle user registration request
- Modify the signup function in RegistrationPageCtrl.java to
use a NewUserView object to get the user's name, email, and
subscriptions.
- Add Mockito to make the email server a mock for all tests.
- Perform integration testing for CacheSqlApiUserService.java.

AngelinaMartineau committed:
Add front end functionality for subscriptions

- add check boxes for each available e-mail subscription
- center the components of the API user registration form
- modify signup function in public_app.js to send chosen e-mail
  subscriptions to the registration page controller
- add data to for availableSUbscriptions to public_app.js

refs #12330

AngelinaMartineau committed:
Add User Subscription Management Page

- add front end functionality for users to update/change their subscriptions
- create ApiUserEmailSubscriptionCtrl.java to handle requests
- add a setSubscriptions function to SqlApiUserDao.java
- add a new query to delete all of a user's subscriptions
- add functionality for a user's current subscriptions to be checked when
the page loads
- add an error page for an invalid api user key

refs #12330

AngelinaMartineau committed:
Add admin utility to send Emails to ApiUsers

- add SQL query to get list of ApiUsers by
  subscription
- add utility to send Mime Multipart emails
  to subscribe ApiUsers
- create AdminApiEmailCtrl.java to handle
  batch email request
- allow users to recieve both plain text and
  text/html versions of emails
- add link to each batch email that sends the
  ApiUser to the subscription management page

refs #12330

AngelinaMartineau committed:
Add Batch Email Front End Functionality

- add email.jsp and email.js
- create page for admins to send batch emails
- add validation check for input from admin
- add dialog box for comfirmation from admin to send
- add preview to let admins view their email in HTML
before sending
- add loading page that displays while emails are sending
- add confirmation page that displays once emails are sent

refs #12330

AngelinaMartineau committed:
Add admins as recipients on batch emails

- refactor sendMessage() in ApiUserBatchEmailServiceImpl.java to
  use two helper functions
- add loop to go through admins in sendMessage() and send them the
  batch email
- add dependency for apache commons EmailValidator to pom.xml
- add a unit test for sendMessage()

refs #12330

AngelinaMartineau committed:
Add unsubscribe link to batch emails

- modify ApiUserBatchEmailServiceImpl.java to include links
  in the email that unsubscribe a user from everything
- add to subscription.js to present a different message when
  a user unsubscribes from all subscriptions directly

refs #12330

AngelinaMartineau committed:
Add option to update user email for subscriptions

- add sql query UPDATE_API_USER_EMAIL
- add function updateEmail to apiUserDao.java
- add unit tests to apiUserDaoIT.java to test
  updateEmail()
- add new mapping to ApiUserEmailSubscriptionCtrl.java
- add to subscription.js and subscriptions.jsp to
  create a new form for updating emails

refs #12330

AngelinaMartineau committed:
Add HTML preview and fix minor bugs

- Add a new tab to batch email box that allows admin
  to view the email they are writing as it will appear
  when sent out
- Modify the error message when a user inputs an email
  already in use for their updated email
- Modify css for the error about an email already existing
- Fix other small bugs

refs #12330

AngelinaMartineau committed:
Add signature to batch emails and add test mode

- Add the option to add a signature to batch emails on
  the admin batch email page
- Add 'Test Mode' to help prevent accidental batch emails
- Add ability for admin to turn test mode on and off
- Modify dialog box on send to mention which mode the
  admin is in.
- Modify back-end so the email is only sent to the admin
  when in test mode
- Add new sendTestMessage function to ApiUserBatchEmailService.java

refs #12330

stouffers commented on src/main/java/gov/nysenate/openleg/controller/api/admin/AdminApiEmailCtrl.java:
```diff
@@ -0,0 +1,53 @@
+    ApiUserBatchEmailServiceImpl apiUserBatchEmailService;
+
+
+    @RequiresPermissions("admin:email:post")
```
The Ctrl files could use some javadocs.  See other Ctrl files for examples of how we document our api controllers.

stouffers commented on src/main/java/gov/nysenate/openleg/controller/api/apiuser/ApiUserEmailSubscriptionCtrl.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.web.bind.annotation.*;
+
+import java.util.*;
+
```
Needs javadocs.  See other Ctrl files for examples

stouffers commented on src/main/java/gov/nysenate/openleg/controller/api/apiuser/ApiUserEmailSubscriptionCtrl.java:
```diff
@@ -0,0 +1,57 @@
+    public void updateSubscriptions(@RequestParam String key, @RequestBody List<String> body) {
+        Set<ApiUserSubscriptionType> subscriptions = new HashSet<>();
+        for (String sub : body) {
+            subscriptions.add(ApiUserSubscriptionType.valueOf(sub));
```
We have a built in method for enum validation that handles cases where invalid values are passed in.
See BaseCtrl#getEnumParameter

stouffers reviewed (changes requested):
Looks good!  Just a few things that should be addressed

AngelinaMartineau committed:
Pull Request Changes

- Add JavaDocs to Registration and Email controllers
- Modify controllers so they user getEnumParameter()
  function
- Modify registerNewUser() function in CachedSqlApiUserService.java
  to take a set of ApiUserSubscriptionTypes rather than strings
- Updated CachedSqlApiUserServiceIT.java to test registerNewUser
  with the correct parameters

refs #12330

AngelinaMartineau committed:
Change Enum Syntax in Test Files

- Modified test files to create ApiUserSubscriptionType
  enums using dot operator

refs #12330

AngelinaMartineau committed:
Change usage of getEnumParameter

- Change the call to getEnumParameter to avoid a
  nullPointerException

refs # 12330

stouffers reviewed (approved):
Looks good
